github.com/cockroachdb/cockroach/pkg/storage: TestReportUnreachableHeartbeats failed under stress

SHA: https://github.com/cockroachdb/cockroach/commits/9deff8d04e4349f27cb8b7a8468ff849c89aa768

Parameters:
```
TAGS=
GOFLAGS=
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=667587&tab=buildLog

```

I180519 10:41:56.469734 66401 storage/store.go:1443  [s1] [n1,s1]: failed initial metrics computation: [n1,s1]: system config not yet available
I180519 10:41:56.469797 66401 gossip/gossip.go:333  [n1] NodeDescriptor set to node_id:1 address:<network_field:"tcp" address_field:"127.0.0.1:41955" > attrs:<> locality:<> ServerVersion:<major_val:0 minor_val:0 patch:0 unstable:0 > 
I180519 10:41:56.479546 66401 storage/engine/rocksdb.go:565  opening rocksdb instance at "/go/src/github.com/cockroachdb/cockroach/artifacts/TestReportUnreachableHeartbeats828426393"
W180519 10:41:56.498078 66401 gossip/gossip.go:1293  [n2] no incoming or outgoing connections
I180519 10:41:56.498266 66256 gossip/client.go:129  [n2] started gossip client to 127.0.0.1:41955
I180519 10:41:56.500020 66401 gossip/gossip.go:333  [n2] NodeDescriptor set to node_id:2 address:<network_field:"tcp" address_field:"127.0.0.1:35793" > attrs:<> locality:<> ServerVersion:<major_val:0 minor_val:0 patch:0 unstable:0 > 
I180519 10:41:56.507565 66401 storage/engine/rocksdb.go:565  opening rocksdb instance at "/go/src/github.com/cockroachdb/cockroach/artifacts/TestReportUnreachableHeartbeats845593380"
W180519 10:41:56.525159 66401 gossip/gossip.go:1293  [n3] no incoming or outgoing connections
I180519 10:41:56.525324 66660 gossip/client.go:129  [n3] started gossip client to 127.0.0.1:41955
I180519 10:41:56.527323 66401 gossip/gossip.go:333  [n3] NodeDescriptor set to node_id:3 address:<network_field:"tcp" address_field:"127.0.0.1:39389" > attrs:<> locality:<> ServerVersion:<major_val:0 minor_val:0 patch:0 unstable:0 > 
I180519 10:41:56.537212 66798 storage/replica_raftstorage.go:520  [s1,r1/1:/M{in-ax}] generated preemptive snapshot f2f0976f at index 15
I180519 10:41:56.537788 66798 storage/store_snapshot.go:605  [s1,r1/1:/M{in-ax}] streamed snapshot to (n2,s2):?: kv pairs: 47, log entries: 5, rate-limit: 2.0 MiB/sec, 1ms
I180519 10:41:56.538055 66690 storage/replica_raftstorage.go:730  [s2,r1/?:{-}] applying preemptive snapshot at index 15 (id=f2f0976f, encoded size=8151, 1 rocksdb batches, 5 log entries)
I180519 10:41:56.540820 66690 storage/replica_raftstorage.go:736  [s2,r1/?:/M{in-ax}] applied preemptive snapshot in 3ms [clear=0ms batch=0ms entries=0ms commit=2ms]
I180519 10:41:56.541340 66798 storage/replica_command.go:1777  [s1,r1/1:/M{in-ax}] change replicas (ADD_REPLICA (n2,s2):2): read existing descriptor r1:/M{in-ax} [(n1,s1):1, next=2]
I180519 10:41:56.544091 66654 storage/replica.go:3312  [txn=cccac695,s1,r1/1:/M{in-ax}] proposing ADD_REPLICA((n2,s2):2): updated=[(n1,s1):1 (n2,s2):2] next=3
I180519 10:41:56.547021 66778 storage/replica_raftstorage.go:520  [s1,r1/1:/M{in-ax}] generated preemptive snapshot 6a0fc8f5 at index 17
I180519 10:41:56.547592 66778 storage/store_snapshot.go:605  [s1,r1/1:/M{in-ax}] streamed snapshot to (n3,s3):?: kv pairs: 50, log entries: 7, rate-limit: 2.0 MiB/sec, 0ms
I180519 10:41:56.547977 66657 storage/replica_raftstorage.go:730  [s3,r1/?:{-}] applying preemptive snapshot at index 17 (id=6a0fc8f5, encoded size=9084, 1 rocksdb batches, 7 log entries)
I180519 10:41:56.551537 66657 storage/replica_raftstorage.go:736  [s3,r1/?:/M{in-ax}] applied preemptive snapshot in 4ms [clear=0ms batch=0ms entries=0ms commit=3ms]
I180519 10:41:56.551991 66778 storage/replica_command.go:1777  [s1,r1/1:/M{in-ax}] change replicas (ADD_REPLICA (n3,s3):3): read existing descriptor r1:/M{in-ax} [(n1,s1):1, (n2,s2):2, next=3]
I180519 10:41:56.553156 65552 storage/raft_transport.go:459  raft transport stream to node 1 established
I180519 10:41:56.557824 66695 storage/replica.go:3312  [txn=b7d2c539,s1,r1/1:/M{in-ax}] proposing ADD_REPLICA((n3,s3):3): updated=[(n1,s1):1 (n2,s2):2 (n3,s3):3] next=4
I180519 10:41:57.374154 66525 storage/node_liveness.go:773  [hb] retrying liveness update after storage.errRetryLiveness: result is ambiguous (context finished during distsender send: context deadline exceeded)
W180519 10:41:57.374191 66525 storage/node_liveness.go:465  [hb] failed node liveness heartbeat: context deadline exceeded
I180519 10:41:57.400437 66638 storage/node_liveness.go:773  [hb] retrying liveness update after storage.errRetryLiveness: result is ambiguous (context finished during distsender send: context deadline exceeded)
W180519 10:41:57.400471 66638 storage/node_liveness.go:465  [hb] failed node liveness heartbeat: context deadline exceeded
I180519 10:41:57.427838 66789 storage/node_liveness.go:773  [hb] retrying liveness update after storage.errRetryLiveness: result is ambiguous (context finished during distsender send: context deadline exceeded)
W180519 10:41:57.427879 66789 storage/node_liveness.go:465  [hb] failed node liveness heartbeat: context deadline exceeded
I180519 10:41:57.636784 66831 util/stop/stopper.go:471  quiescing; tasks left:
1      [async] kv.TxnCoordSender: aborting txn
I180519 10:41:57.636817 66829 util/stop/stopper.go:471  quiescing; tasks left:
6      [async] storage.multiTestContextKVTransport: calling next replica
1      [async] kv.TxnCoordSender: aborting txn
I180519 10:41:57.636862 66829 util/stop/stopper.go:471  quiescing; tasks left:
5      [async] storage.multiTestContextKVTransport: calling next replica
1      [async] kv.TxnCoordSender: aborting txn
I180519 10:41:57.636885 66830 util/stop/stopper.go:471  quiescing; tasks left:
1      [async] kv.TxnCoordSender: aborting txn
I180519 10:41:57.636917 66829 util/stop/stopper.go:471  quiescing; tasks left:
4      [async] storage.multiTestContextKVTransport: calling next replica
1      [async] kv.TxnCoordSender: aborting txn
I180519 10:41:57.636939 66829 util/stop/stopper.go:471  quiescing; tasks left:
3      [async] storage.multiTestContextKVTransport: calling next replica
1      [async] kv.TxnCoordSender: aborting txn
I180519 10:41:57.636962 66829 util/stop/stopper.go:471  quiescing; tasks left:
2      [async] storage.multiTestContextKVTransport: calling next replica
1      [async] kv.TxnCoordSender: aborting txn
I180519 10:41:57.636990 66829 util/stop/stopper.go:471  quiescing; tasks left:
1      [async] storage.multiTestContextKVTransport: calling next replica
1      [async] kv.TxnCoordSender: aborting txn
I180519 10:41:57.637017 66829 util/stop/stopper.go:471  quiescing; tasks left:
1      [async] kv.TxnCoordSender: aborting txn
W180519 10:41:57.637243 66789 storage/node_liveness.go:465  [hb] failed node liveness heartbeat: node unavailable; try another peer
W180519 10:41:57.637410 66638 storage/node_liveness.go:465  [hb] failed node liveness heartbeat: node unavailable; try another peer
W180519 10:41:57.637570 66525 storage/node_liveness.go:465  [hb] failed node liveness heartbeat: node unavailable; try another peer
I180519 10:41:57.639997 66828 storage/engine/rocksdb.go:698  closing rocksdb instance at "/go/src/github.com/cockroachdb/cockroach/artifacts/TestReportUnreachableHeartbeats190097314"
I180519 10:41:57.640271 66828 storage/engine/rocksdb.go:698  closing rocksdb instance at "/go/src/github.com/cockroachdb/cockroach/artifacts/TestReportUnreachableHeartbeats828426393"
I180519 10:41:57.641767 66828 storage/engine/rocksdb.go:698  closing rocksdb instance at "/go/src/github.com/cockroachdb/cockroach/artifacts/TestReportUnreachableHeartbeats845593380"
panic: runtime error: index out of range [recovered]
	panic: runtime error: index out of range [recovered]
	panic: runtime error: index out of range

goroutine 66401 [running]:
testing.tRunner.func1(0xc421212d20)
	/usr/local/go/src/testing/testing.go:742 +0x29d
panic(0x20cee40, 0x3493b10)
	/usr/local/go/src/runtime/panic.go:505 +0x229
github.com/cockroachdb/cockroach/pkg/util/leaktest.AfterTest.func1()
	/go/src/github.com/cockroachdb/cockroach/pkg/util/leaktest/leaktest.go:89 +0x432
panic(0x20cee40, 0x3493b10)
	/usr/local/go/src/runtime/panic.go:505 +0x229
github.com/cockroachdb/cockroach/pkg/storage_test.TestReportUnreachableHeartbeats(0xc421212d20)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/client_raft_test.go:2042 +0x81a
testing.tRunner(0xc421212d20, 0x248c6c8)
	/usr/local/go/src/testing/testing.go:777 +0xd0
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:824 +0x2e0

goroutine 1 [chan receive]:
testing.(*T).Run(0xc421212d20, 0x238b8e7, 0x1f, 0x248c6c8, 0x884201)
	/usr/local/go/src/testing/testing.go:825 +0x301
testing.runTests.func1(0xc420538870)
	/usr/local/go/src/testing/testing.go:1063 +0x64
testing.tRunner(0xc420538870, 0xc420767b90)
	/usr/local/go/src/testing/testing.go:777 +0xd0
testing.runTests(0xc420512100, 0x34bdc00, 0x1f1, 0x1f1, 0x213a080)
	/usr/local/go/src/testing/testing.go:1061 +0x2c4
testing.(*M).Run(0xc420082100, 0x0)
	/usr/local/go/src/testing/testing.go:978 +0x171
github.com/cockroachdb/cockroach/pkg/storage_test.TestMain(0xc420082100)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/main_test.go:57 +0x1e3
main.main()
	_testmain.go:1050 +0x151

goroutine 20 [syscall]:
os/signal.signal_recv(0x0)
	/usr/local/go/src/runtime/sigqueue.go:139 +0xa6
os/signal.loop()
	/usr/local/go/src/os/signal/signal_unix.go:22 +0x22
created by os/signal.init.0
	/usr/local/go/src/os/signal/signal_unix.go:28 +0x41

goroutine 41 [chan receive]:
github.com/cockroachdb/cockroach/pkg/util/log.flushDaemon()
	/go/src/github.com/cockroachdb/cockroach/pkg/util/log/clog.go:1158 +0xf1
created by github.com/cockroachdb/cockroach/pkg/util/log.init.0
	/go/src/github.com/cockroachdb/cockroach/pkg/util/log/clog.go:591 +0x110

goroutine 42 [chan receive]:
github.com/cockroachdb/cockroach/pkg/util/log.signalFlusher()
	/go/src/github.com/cockroachdb/cockroach/pkg/util/log/clog.go:598 +0xab
created by github.com/cockroachdb/cockroach/pkg/util/log.init.0
	/go/src/github.com/cockroachdb/cockroach/pkg/util/log/clog.go:592 +0x128

goroutine 43 [select, locked to thread]:
runtime.gopark(0x248e8f0, 0x0, 0x235ca46, 0x6, 0x18, 0x1)
	/usr/local/go/src/runtime/proc.go:291 +0x11a
runtime.selectgo(0xc42006bf50, 0xc420608ae0)
	/usr/local/go/src/runtime/select.go:392 +0xe50
runtime.ensureSigM.func1()
	/usr/local/go/src/runtime/signal_unix.go:549 +0x1f4
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:2361 +0x1

goroutine 39970 [chan receive]:
github.com/cockroachdb/cockroach/pkg/util/log.(*loggingT).gcDaemon(0xc42185a000)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/log/clog.go:1210 +0x68
created by github.com/cockroachdb/cockroach/pkg/util/log.NewSecondaryLogger
	/go/src/github.com/cockroachdb/cockroach/pkg/util/log/secondary_log.go:82 +0x2ab

goroutine 39971 [chan receive]:
github.com/cockroachdb/cockroach/pkg/util/log.(*loggingT).gcDaemon(0xc42185a120)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/log/clog.go:1210 +0x68
created by github.com/cockroachdb/cockroach/pkg/util/log.NewSecondaryLogger
	/go/src/github.com/cockroachdb/cockroach/pkg/util/log/secondary_log.go:82 +0x2ab


ERROR: exit status 2

context canceled
make: *** [stress] Error 1
```

--- a/pkg/storage/client_raft_test.go
+++ b/pkg/storage/client_raft_test.go
@@ -2033,9 +2033,17 @@ func TestReportUnreachableHeartbeats(t *testing.T) {
 	mtc.replicateRange(rangeID, 1, 2)
 
 	leaderIdx := -1
-	for i, store := range mtc.stores {
-		if store.RaftStatus(rangeID).SoftState.RaftState == raft.StateLeader {
-			leaderIdx = i
+	// Loop until a leader is elected.
+	for {
+		for i, store := range mtc.stores {
+			if store.RaftStatus(rangeID).SoftState.RaftState == raft.StateLeader {
+				leaderIdx = i
+				break
+			}
+		}
+		if leaderIdx == -1 {
+			runtime.Gosched()
+		} else {
 			break
 		}
 	}

